internal/download/get.go

```diff
@@ -31,81 +34,121 @@ func init() {
+	if err != nil {
+		logging.Debug("Could not grab url: %v", err)
+	} else if res.StatusCode != http.StatusOK {
+		logging.Debug("Could not grab url due to statuscode: %d", res.StatusCode)
 	} else {
-		total, err = strconv.Atoi(length)
+		lengthInt, err := strconv.Atoi(res.Header.Get("Content-Length"))
 		if err != nil {
-			logging.Debug("Content-length: %v", length)
-			return nil, failures.FailInput.Wrap(err)
+			logging.Debug("Could not grab content-length: %v", err)
+		} else {
+			length = int64(lengthInt)
 		}
 	}
 
-	bar := progress.AddByteProgressBar(int64(total))
+	// Close early cause we're just looking at the header
+	// Yes normally you'd use a HEAD for this, but S3 presigned URLs don't support HEAD requests
+	res.Body.Close()
 
-	src := resp.Body
-	var dst bytes.Buffer
+	// Record progress
+	bar := progress.AddByteProgressBar(length)
+	defer bar.Abort(true) // ensure we don't get stuck on an incomplete bar
 
-	src = bar.ProxyReader(resp.Body)
+	cb := func(length int) {
+		if !bar.Completed() {
+			// Failsafe, so we don't get blocked by a progressbar
+			bar.IncrBy(length)
+		}
+	}
 
-	_, err = io.Copy(&dst, src)
+	// Prepare result recorder
+	b := []byte{}
+	w := NewWriteAtBuffer(b, cb)
+
+	dl := s3manager.NewDownloader(sess)
+	dl.RequestOptions = append(dl.RequestOptions, func(r *request.Request) {
+		r.Handlers.Build.PushBack(func(r *request.Request) {
+			// Work around AWS rewriting our query in the wrong order, causing signing to fail
+			r.HTTPRequest.URL.RawQuery = url.RawQuery
+		})
+	})
+	_, err = dl.Download(w,
+		&s3.GetObjectInput{
+			Bucket: aws.String(s3m.Bucket),
+			Key:    aws.String(s3m.Key),
+		})
 	if err != nil {
-		return nil, failures.FailInput.Wrap(err)
+		return nil, locale.WrapError(err, "err_dl_s3", "Downloading failed due to underlying S3 error: {{.V0}}.", err.Error())
 	}
 
-	if !bar.Completed() {
-		// Failsafe, so we don't get blocked by a progressbar
-		bar.IncrBy(total)
-	}
+	return w.Bytes(), nil
+}
 
-	return dst.Bytes(), nil
+type s3Meta struct {
+	Bucket string
+	Region string
+	Key    string
 }
 
-func _testHTTPGetWithProgress(url string, progress *progress.Progress) ([]byte, *failures.Failure) {
-	return _testHTTPGet(url)
+func parseS3URL(url *url.URL) (s3Meta, error) {
+	r := s3Meta{Key: url.Path}
+	domain := strings.SplitN(url.Host, ".", 4)
+	if len(domain) != 4 {
+		return r, locale.NewError("err_s3_host", "API responded with an invalid artifact host: {{.V0}}.", url.Host)
+	}
+	if strings.HasSuffix(url.Host, ".s3.amazonaws.com") { // https://bucket-name.s3.amazonaws.com/key-name
+		r.Bucket = domain[0]
+		r.Region = constants.DefaultS3Region
+	} else if domain[1] == "s3" && domain[3] == "amazonaws" { // https://bucket-name.s3.Region.amazonaws.com/key-name
+		r.Bucket = domain[0]
+		r.Region = domain[2]
+	} else { // https://s3.Region.amazonaws.com/bucket-name/key-name
+		r.Bucket = strings.SplitN(url.Path, "/", 1)[0]
+		r.Region = domain[1]
+	}
+	return r, nil
 }
```

Comment:
This appears to be broken now: https://play.golang.org/p/11uiIA8xTON

Reply:
I don't see a problem?

Reply:
- https://bucket-name.s3.Region.amazonaws.com/key-name/extra
  - bucket: "/key-name/extra" region: "s3" key: "/key-name/extra"
    - bucket and region are incorrect
- https://bucket-name.s3.amazonaws.com/key-name/extra
  - bucket: "bucket-name" region: "us-east-1" key: "/key-name/extra"
    - looks good
- https://s3.Region.amazonaws.com/bucket-name/key-name/extra
  - bucket: "/bucket-name/key-name/extra" region: "Region" key: "/bucket-name/key-name/extra"
    - bucket and key are incorrect

Also, my version strips the slash prefix, but I'm not sure if that's needed.